azurerm/resource_arm_private_link_service.go

```diff
@@ -0,0 +1,649 @@
+				PrivateEndpoint:                   expandArmPrivateLinkServicePrivateEndpoint(privateEndpoint),
+				PrivateLinkServiceConnectionState: expandArmPrivateLinkServicePrivateLinkServiceConnectionState(privateLinkServiceConnectionState),
+			},
+		}
+
+		results = append(results, result)
+	}
+	return &results
+}
+
+func expandArmPrivateLinkServicePrivateLinkServicePropertiesVisibility(input []interface{}) *network.PrivateLinkServicePropertiesVisibility {
+	if len(input) == 0 {
+		return nil
+	}
+
+	subscriptions := make([]string, 0)
+
+	for _, v := range input {
+		subscriptions = append(subscriptions, v.(string))
+	}
+
+	result := network.PrivateLinkServicePropertiesVisibility{
+		Subscriptions: &subscriptions,
+	}
+
+	return &result
+}
+
+func expandArmPrivateLinkServicePrivateEndpoint(input []interface{}) *network.PrivateEndpoint {
+	if len(input) == 0 {
+		return nil
+	}
+	v := input[0].(map[string]interface{})
+
+	id := v["id"].(string)
+	location := azure.NormalizeLocation(v["location"].(string))
+	t := v["tags"].(map[string]interface{})
+
+	result := network.PrivateEndpoint{
+		ID:       utils.String(id),
+		Location: utils.String(location),
+		Tags:     tags.Expand(t),
+	}
+	return &result
+}
+
+func expandArmPrivateLinkServicePrivateLinkServiceConnectionState(input []interface{}) *network.PrivateLinkServiceConnectionState {
+	if len(input) == 0 {
+		return nil
+	}
+	v := input[0].(map[string]interface{})
+
+	status := v["status"].(string)
+	description := v["description"].(string)
+	actionRequired := v["action_required"].(string)
+
+	result := network.PrivateLinkServiceConnectionState{
+		ActionRequired: utils.String(actionRequired),
+		Description:    utils.String(description),
+		Status:         utils.String(status),
+	}
+	return &result
+}
+
+func flattenArmPrivateLinkServicePrivateLinkServicePropertiesAutoApproval(input *network.PrivateLinkServicePropertiesAutoApproval) []string {
+	result := make([]string, 0)
+	if input == nil {
+		return result
+	}
+
+	for _, v := range *input.Subscriptions {
+		if subscription := v; subscription != "" {
+			result = append(result, subscription)
+		}
+	}
+
+	return result
+}
+
+func flattenArmPrivateLinkServicePrivateLinkServiceIPConfiguration(input *[]network.PrivateLinkServiceIPConfiguration) []interface{} {
+	results := make([]interface{}, 0)
+	if input == nil {
+		return results
+	}
+
+	for _, item := range *input {
+		v := make(map[string]interface{})
+
+		if name := item.Name; name != nil {
+			v["name"] = *name
+		}
+		if privateLinkServiceIPConfigurationProperties := item.PrivateLinkServiceIPConfigurationProperties; privateLinkServiceIPConfigurationProperties != nil {
+			v["private_ip_allocation_method"] = string(privateLinkServiceIPConfigurationProperties.PrivateIPAllocationMethod)
+			if privateIpAddress := privateLinkServiceIPConfigurationProperties.PrivateIPAddress; privateIpAddress != nil {
+				v["private_ip_address"] = *privateIpAddress
+			}
```

Comment:
these are easier to review & read when
```suggestion
			if v := privateLinkServiceIPConfigurationProperties.PrivateIPAddress; v != nil {
				v["private_ip_address"] = *v
			}
```

Reply:
fixed.